go-controller/pkg/node/helper_linux.go

```diff
@@ -15,27 +16,62 @@ import (
 
 // getDefaultGatewayInterfaceDetails returns the interface name on
 // which the default gateway (for route to 0.0.0.0) is configured.
-// It also returns the default gateway itself.
-func getDefaultGatewayInterfaceDetails() (string, net.IP, error) {
-	routes, err := netlink.RouteList(nil, syscall.AF_INET)
+// It also returns the default gateways themselves.
+func getDefaultGatewayInterfaceDetails() (string, []net.IP, error) {
+	var family int
+	var intfName string
+	var gatewayIPs []net.IP
+	var needIPv4 bool
+	var needIPv6 bool
+
+	if config.IPv4Mode {
+		family = syscall.AF_INET
+		needIPv4 = true
+	}
+	if config.IPv6Mode {
+		family = syscall.AF_INET6
+		needIPv6 = true
+	}
+	if config.IPv4Mode && config.IPv6Mode {
+		family = syscall.AF_UNSPEC
+	}
```

Comment:
You can always use `AF_UNSPEC`; the wrong-family addresses will just get ignored anyway because, eg, `needIPv6` isn't set. So all this is just

    needIPv4 := config.IPv4Mode
    needIPv6 := config.IPv6Mode